### PR TITLE
EnumerableSet: Clear

### PR DESCRIPTION
**Pull request by wjmelements**
#### Changes
It should be cheaper to remove all items.
It won't always be possible to do this because of the block gas limit, but when it is this should be much cheaper than sequential removal.
Reviewers @nventuro

### CODE REVIEW

**Comment by wjmelements** [comment]
```
TypeError: Member "length" is read-only and cannot be used to resize arrays.
  --> contracts/utils/structs/EnumerableSet.sol:71:9:
   |
71 |         set._values.length = 0;
   |         ^^^^^^^^^^^^^^^^^^
```
I hate that they removed this functionality. It was so good; it zeroed all of the elements too.

**Comment by wjmelements** [comment]
If this feature will be accepted, I will write tests for it.

**Comment by Amxx** [comment]
Hello @wjmelements, and that you for your many PR addressing gas usage! While we didn't found ourselves in a situation where this feature was required, it would be a nice to have that user might find handy.

I'm worried about the "loop" part, and the fact that a smart contract relying on this feature might be in trouble if the set includes too many elements for the clear to git in one transaction.

> ```
> TypeError: Member "length" is read-only and cannot be used to resize arrays.
>   --> contracts/utils/structs/EnumerableSet.sol:71:9:
>    |
> 71 |         set._values.length = 0;
>    |         ^^^^^^^^^^^^^^^^^^
> ```
> 
> I hate that they removed this functionality. It was so good; it zeroed all of the elements too.

While zeroing the elements is a nice addition, I don't think it is strictly necessary. Setting the length to 0, even without the elements being erased, should maintain the safe behavior of the set. You would not get the refund associated with erasing the elements, but then putting new data in will also be cheaper.
While solidity prevents it, I do believe it can still be done in assembly. What do you think?

**Comment by wjmelements** [comment]
> While we didn't found ourselves in a situation where this feature was required, it would be a nice to have that user might find handy.

https://github.com/airswap/airswap-protocols/blob/7a6ca7aa1d22e4c335b17e54a506d634fd177a6d/source/registry/contracts/Registry.sol#L64

> relying on this feature might be in trouble if the set includes too many elements for the clear to git in one transaction.

In my case there is [another way to remove individual elements](https://github.com/airswap/airswap-protocols/blob/7a6ca7aa1d22e4c335b17e54a506d634fd177a6d/source/registry/contracts/Registry.sol#L48) that we can fall back to. Can recommend that via docs.

> You would not get the refund associated with erasing the elements, but then putting new data in will also be cheaper.
> While solidity prevents it, I do believe it can still be done in assembly. What do you think?

Fine by me, especially if they go forward with [removing the refund incentive in London](https://eips.ethereum.org/EIPS/eip-3403).

**Comment by wjmelements** [comment]
> While solidity prevents it, I do believe it can still be done in assembly. What do you think?

I started this but I am getting an error I don't see a way to disable. I've looked through the solidity source code and found no such flag. The sage advice on the stack exchange is to rename the functions that collide. I suggest renaming `add` to `insert`. @Amxx does that sound good to you?

```
Warning: Variable is shadowed in inline assembly by an instruction of the same name
   --> contracts/utils/structs/EnumerableSet.sol:159:5:
    |
159 |     function add(Bytes32Set storage set, bytes32 value) internal returns (bool) {
    |     ^ (Relevant source part starts here and spans across multiple lines).


Warning: Variable is shadowed in inline assembly by an instruction of the same name
   --> contracts/utils/structs/EnumerableSet.sol:217:5:
    |
217 |     function add(AddressSet storage set, address value) internal returns (bool) {
    |     ^ (Relevant source part starts here and spans across multiple lines).


Warning: Variable is shadowed in inline assembly by an instruction of the same name
   --> contracts/utils/structs/EnumerableSet.sol:275:5:
    |
275 |     function add(UintSet storage set, uint256 value) internal returns (bool) {
    |     ^ (Relevant source part starts here and spans across multiple lines).
```

**Comment by Amxx** [comment]
We should definitely NOT rename any functions. This would be a braking change in the library and we want to avoid that at all cost.

Also, this warning was removed since solc 0.8.2. Can you confirm by running the code against 0.8.2 or 0.8.3 ?

**Comment by wjmelements** [comment]
> We should definitely NOT rename any functions. This would be a breaking change in the library and we want to avoid that at all cost.
> 
> Also, this warning was removed since solc 0.8.2. Can you confirm by running the code against 0.8.2 or 0.8.3 ?

Yes, that does address the warning. Should I commit the change in `hardhat.config.js` or leave it as-is?

**Comment by wjmelements** [comment]
> Yes, that does address the warning. Should I commit the change in `hardhat.config.js` or leave it as-is?

@Amxx My intuition is that I should leave it as-is because it doesn't break any of the checks and other patch-related issues may go unnoticed despite the intended compiler versions to support, but the counter-points are that there are no other build warnings, and end-users may see the warning and file issues.

**Comment by wjmelements** [comment]
I see the solc version is bumped in https://github.com/OpenZeppelin/openzeppelin-contracts/pull/2612 so I will merge master after that merges.

File: contracts/utils/structs/EnumerableSet.sol [hunk 1]
```diff
@@ -63,6 +63,26 @@ library EnumerableSet {
...
+    /**
+     * @dev Removes all elements from the set. O(n).
+     *
+     * Note that this can run out of gas for large sets.
+     * First use `remove` to reduce the size of a large set in this case.
+     */
+    function _clear(Set storage set) private {
+        uint256 last = set._values.length;
+        unchecked {
+            while (last --> 0) {
```
Comment (eMarchenko) [inline contracts/utils/structs/EnumerableSet.sol:75]: tbh, I've spent 15 minutes figuring out what is this `-->` syntax. Eventually my colleague saved me as he guessed it is `last-- > 0`. I wasn't even aware that it is possible. While this language feature is cool and I really like the way it looks, it seems to be too advanced for the most popular contract library.

Comment (wjmelements) [reply contracts/utils/structs/EnumerableSet.sol:75]: This is the best way to loop to zero inclusive on an unsigned variable. I've always thought it more clear when written `-->`, but I would rearrange the whitespace upon reviewer request.

**Comment by wjmelements** [comment]
ptal @Amxx

**Comment by frangio** [comment]
We've been discussing this and we don't feel comfortable including function that loops over an array of unbounded length, so we're not going to merge it.

We evaluated using assembly to set the length to zero while leaving the old set values in storage, but this seems to still require a loop over the indices array. It may be possible to get it working with some more tricks, we encourage people to try this out and let us know if they find a clean way to implement set clear without any loops.

A potential way to implement "clearing" is to throw away the set and use a new one, by keeping a mapping (or an array?) of sets, where bumping the index to the currently used set is equivalent to clearing. This will incur an additional storage load.

File: contracts/utils/structs/EnumerableSet.sol (hunk 1 above)
Comment (wjmelements) [reply contracts/utils/structs/EnumerableSet.sol:75]: Additional matter of interest: decrement inside solidity loop conditions doesn't seem to work at all in 0.8.0. I am concerned at how the tests still passed.

**Pull request closed without merging**